bundle/edu.gemini.programexport.servlet/src/main/scala/edu/gemini/programexport/servlet/ProgramExportServlet.scala

```diff
@@ -0,0 +1,209 @@
+package edu.gemini.programexport.servlet
+
+import java.io.BufferedOutputStream
+import java.security.Principal
+import java.util.logging.{Level, Logger}
+import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse}
+import edu.gemini.pot.spdb.IDBDatabaseService
+import edu.gemini.spModel.core.{ProgramId, SPProgramID}
+import scalaz._
+import Scalaz._
+import edu.gemini.pot.sp.{ISPGroup, ISPNode, ISPObsComponent, ISPObservation, ISPProgram, SPComponentType}
+import edu.gemini.shared.util.immutable.ScalaConverters._
+import edu.gemini.spModel.data.YesNoType
+import edu.gemini.spModel.gemini.obscomp.{SPProgram, SPSiteQuality}
+import edu.gemini.spModel.obs.SPObservation
+import edu.gemini.spModel.obscomp.{ProgramNote, SPGroup, SPNote, SchedNote}
+import edu.gemini.spModel.target.env.TargetEnvironment
+import edu.gemini.spModel.util.SPTreeUtil
+
+import javax.servlet.ServletException
+import scala.collection.JavaConverters._
+
+final case class ProgramExportServlet(odb: IDBDatabaseService, user: Set[Principal]) extends HttpServlet {
+
+  import ProgramExportServlet._
+
+  override def doPost(request: HttpServletRequest, response: HttpServletResponse): Unit =
+    doRequest(request, response)
+
+  override def doGet(request: HttpServletRequest, response: HttpServletResponse): Unit =
+    doRequest(request, response)
+
+  def doRequest(request: HttpServletRequest, response: HttpServletResponse): Unit =
+  // Attempt to retrieve the program name: ProgramID instead?
+    ParamParser(request).id.fold(t => throw new ServletException(t), build(_, response))
+
+  // Build the JSON in a bottom-up approach starting at ISPProgram.
+  // Traverse a node's children and on the way back up, add the information for the node.
+  // Right now we just return an empty String: whether to use Strings or Argonaut to build the
+  // JSON is a decision I'm leaving to the end.
```

Comment:
Something along these lines (which I haven't compiled :-) )

```scala
import argonaut.Json
import argonaut.Json.JsonAssoc

def process(n: ISPNode): Option[JsonAssoc] =

  n match {
    case _: ISPProgram | _: ISPGroup | _: ISPObservation | _: ISPObsComponent | _: ISPQaLog =>
      simpleNode(n)

    case _: ISPSeqComponent =>
      sequenceNode(n)

    case _ =>
      None
 
  }


def simpleNode(n: ISPNode): Option[JsonAssoc] =
  
  for {
    t <- n.dataObject.map(_.getType)
    o <- componentFields(n)
  } yield (
    t.name,
    n.children.foldLeft(("key" -> n.getNodeKey.toString) ->: o) { (oʹʹ, c) =>
      process(c) ->?: o''
    }
  )
  
  
def componentFields(n: ISPNode): Option[Json] =

  n.dataObject.flatMap {
    case p: SPProgram     => Some(programFields(p))
    case g: SPGroup       => Some(groupFields(g))
    case o: SPObservation => Some(observationFields(o))

    case s: SPSiteQuality => Some(siteQualityFields(s))
    // etc. etc.

    case _                => None

  }

def programFields(p: SPProgram): Json = 
  Json.jEmptyObject                        :+
    ("isThesis" -> Json.jBool(p.isThesis)) :+
    ...

def groupFields(g: SPGroup): Json =
  ...


def sequenceNode(n: ISPNode): Option[JsonAssoc] =

  n.dataObject.flatMap {
    case _: SeqBase => Option(n.getContextObservatgion).map(sequence)
    case _          => None
  }


def sequence(o: ISPObservation): JsonAssoc = {
   val steps = ConfigBridge.extractSequence(o, null, ConfigValMapInstances.TO_SEQUENCE_VALUE).getAllSteps.toList

   "sequence" -> Json.jArray(steps.map { s =>
     s.itemEntries.toList.foldLeft(Json.jEmptyObject) { (o, e) =>
       (e.getKey.getPath -> e.getItemValue) ->: o   // e is an ItemEntry and the value will have been mapped to a String by TO_SEQUENCE_VALUE
     }
   })
}
```

Reply:
Oh, you'd need to import `edu.gemini.spModel.rich.pot.sp._` as well.

Reply:
This really helps, especially considering I haven't used Argonaut before, so it'll save me a ton of time digging through old code or docs.

Maybe it's my sleep addled brain, but what is the "official" distinction between a simple node and a component field?

Reply:
Maybe it isn't a good name.  It's `simpleNode` for want of a better name vs. `sequenceNode` since the two are handled differently.  We don't walk sequence nodes but instead we run the sequence and extract it into an array of JSON objects

`componentFields` correspond to the interesting fields in a component's data object.  That's where you extract the thesis, etc. from `SPProgram` etc.  In `simpleNode` we take the component fields, add a "key" field with the node key, and then add objects for each child that produces a result.

Reply:
Another idea that just occurred to me.  You can turn any data object into a `ParamSet` and it should be straightforward to map a `ParamSet` to a `Json` object.  `ParamSet` is just an intermediate format that we turn into XML.  This would make it so you don't have to match on the various kinds of data objects and pull out stuff individually, though it would pull in extra stuff that wasn't asked for.

Reply:
The ParamSet idea is a good one, but it may not work quite right with regards to things like `Durations` and what not, I don't think.

I'm trying to figure out how to just get everything pretty printed so I can read it at this point and it's not just a blob of text, how to get the acquisition time, and if I have all the nodes I need. Things still don't seem complete and the blob is not readable. I'm just not sure where to put the `print2` or `print4` and it doesn't seem to be registering on the `Json` instance even though I'm importing all of `argonaut` and `Argonaut`.

Reply:
I am convinced that Argonaut is made by people who hate happiness.